File: C#/LinkedList/SinglyLinkedList/LinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace dataStructures
{
    class LinkedListNode<T>
    {
        public T value { get; set; }
        public LinkedListNode<T> next { get; set; }
    };
    class LinkedList<T>
    {
       
        public LinkedListNode<T> head { get;  set; }
        public LinkedListNode<T> tail { get;  set; }

        public LinkedList<T> prepend(T value)
        {
            var node = new LinkedListNode<T> { value = value };

            if(head == null)
            {
                head = node;
                tail = node;
             

            } else
            {
                node.next = head;
                head = node;
            }
            return this;
        }

        public LinkedList<T> append(T value)
        {
            var node = new LinkedListNode<T> { value = value };

            if (head == null)
            {
                head = node;
                tail = node;
            }
            else
            {
                tail.next = node;
                tail = node;
            }
            return this;
        }

        public LinkedList<T> remove(T value)
        {
            var lastNode = head;
            var currentNode = head;

            if(head.value.Equals(value))
            {
                head = head.next;
                return this;
            }


            while (!currentNode.next.Equals(null)) 
            {
                if (object.Equals(value, currentNode.value))
                {
                    lastNode.next = currentNode.next;
                    currentNode.next = null;
                    return this;

                }
                else
                {
                    lastNode = currentNode;
                    currentNode = currentNode.next;
                }
            }

            return this;
        }


    }
}

File: typescript/src/List/LinkedList.ts

class LinkedListNode<T> {
    constructor(
        
        public value: T, 
        // next node
        public next: LinkedListNode<T> = null ){ };
}
export class LinkedList<T>{
    head:LinkedListNode<T> = null;
    tail:LinkedListNode<T> = null;
    count = 0;

    public prepend(value:T){
        this.count++

        // create a linkedList Object
        const listNode = new LinkedListNode(value);
        // no head, when running for the first time
        if(!this.head){
            this.head = listNode;
            this.tail = listNode;
            
            // return the linkedList Object after appending
            return this;
        }

        // if there is already a head, we need to make our node the new head
        listNode.next = this.head;
        this.head = listNode;
        return this;
    }


    public append(value:T){
        this.count++

         // create a linkedList Object
         const listNode = new LinkedListNode(value);
         // no head, when running for the first time
         if(!this.head){
             this.head = listNode;
             this.tail = listNode;
             
             // return the linkedList Object after appending
             return this;
         }

        // if there is a head, we need make our node the new tail
        
        this.tail.next = listNode;
        // keep the tail instance for ref 
        this.tail = listNode;
        return this;

    }

    public remove(value : T): LinkedListNode<T>{
        // console.log(this.tail)
        if(this.head.value == value){
            const toRemoveNode = JSON.parse(JSON.stringify(this.head))

            if(!this.head.next){
                this.head = null;
                this.tail = null
            } else {
                this.head = this.head.next
            }
            this.count--;
            return toRemoveNode;
        }

      

        let currentNode = this.head;
        let lastNode = this.head;
        while(currentNode.value){
            // console.log(currentNode)
            if(currentNode.value != value){
                lastNode = currentNode;
                currentNode = currentNode.next;
                // console.log("lastNode",lastNode)
                // console.log({currentNode}, {lastNode})
            } else {

                // console.log('rere')
                if(this.tail == currentNode){
                    this.tail = lastNode;
                    // console.log('yes', this.tail)
                } else {
                    
                    lastNode.next = currentNode.next;
                    
                }
                // console.log(this.tail)
                const toRemoveNode = JSON.parse(JSON.stringify(currentNode))
                currentNode.next = null;
                currentNode.value = null;
                // console.log(this.tail)
                return toRemoveNode;
            }
            
        }
        return null;


    }


    public find(value : T|string, callbackFunction?: (value: any) => boolean) : LinkedListNode<T>{
        if(!this.head){
            return undefined;
        }
        
        let currentNode = this.head;

        do{  
            if(callbackFunction){
                const node = this.findWithCallback(currentNode, callbackFunction);
                if(node){
                    return node;
                }
            } else {

                if(currentNode.value != value){
                    currentNode = currentNode.next;
                } else {
                    return currentNode;
                }
            }
        } while((currentNode.next));
        return;
    }
    
    private findWithCallback(currentNode : LinkedListNode<T>, callbackFunction:(value: any) => boolean) : LinkedListNode<T>{
       
            if(!callbackFunction(currentNode.value)){
                currentNode = currentNode.next;
            } else {
                return currentNode;
            }
            return null
        
    }
}